lib/spack/spack/util/module_cmd.py

```diff
@@ -162,6 +158,43 @@ def path_from_modules(modules):
         # that we give preference to the last module to be loaded
         # for packages requiring to load multiple modules in sequence
         best_choice = candidate_path or best_choice
+
+    # if we found something return it
+    return best_choice
+
+
+def path_from_modules(modules, hint_name=None):
+    """Inspect a list of TCL modules for entries that indicate the absolute
+    path at which the library supported by said module can be found.
+
+    Args:
+        modules (list): module files to be loaded to get an external package
+        hint_name (str): optional hint giving a name that likely maps
+                         is a better choice than choosing the last
+                         possible path.
+                         A good choice is the package name.
+                         E.g., if operating on an external mvapich,
+                         then give priority to modules with '.*mvapich.*'
+                         in their name. From these, always choose the last
+
+    Returns:
+        Guess of the prefix path where the package
+    """
+    assert isinstance(modules, list), 'the "modules" argument must be a list'
+
+    remaining_modules = modules
+    if hint_name:
+        # first choices are those that have the hint in their names
+        first_choice_modules = [m for m in modules if hint_name in m]
+        best_choice = _path_from_modules_helper(first_choice_modules)
+        # if we found something return it
+        if best_choice:
+            return best_choice
+        # if we didn't succeed, then compute the remaining modules
+        remaining_modules = [m for m in modules if m not in first_choice_modules]
+
+    # search the final set of modules
+    best_choice = _path_from_modules_helper(remaining_modules)
```

Comment:
We can avoid the restriction that modules has to be a list and allow any iterable with a defensive copy.

```suggestion
    modules = list(modules)  # defensive copy for generators
    if hint_name:
        # first choices are those that have the hint in their names
        first_choice_modules = [m for m in modules if hint_name in m]
        best_choice = _path_from_modules_helper(first_choice_modules)
        # if we found something return it
        if best_choice:
            return best_choice
        # if we didn't succeed, then compute the remaining modules
        modules = [m for m in modules if m not in first_choice_modules]

    # search the final set of modules
    best_choice = _path_from_modules_helper(modules)
```

Reply:
That code was pre-existing - I could careless, so long as you can take a subset that, e.g., contains the `hint`, which is usually the package name.

From my experience, my method severally cuts down on the number of modules evaluated - because typically picking the modules that contains the package_name is the right choice. (and yield 1 module). This hasn't seemed to break my CCE toolchain either

Reply:
Oh I see I had missed the one line deletion hiding in the previous method! Doh!